go/runtime/host/sandbox/connector.go

```diff
@@ -0,0 +1,130 @@
+
+	// Connect establishes a connection to the runtime.
+	Connect(p process.Process) (net.Conn, error)
+
+	// Close releases any resources associated with the connector.
+	Close()
+}
+
+// UnixSocketConnector is a runtime connector that uses a UNIX socket to communicate with the
+// runtime.
+type UnixSocketConnector struct {
+	socketPath string
+	sandboxed  bool
+	listener   *net.UnixListener
+	logger     *logging.Logger
+}
+
+// NewUnixSocketConnector returns a new UNIX socket connector.
+func NewUnixSocketConnector(logger *logging.Logger, runtimeDir string, sandboxed bool) (Connector, error) {
+	hostSocket := filepath.Join(runtimeDir, "host.sock")
+	listener, err := net.ListenUnix("unix", &net.UnixAddr{Name: hostSocket})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create host socket: %w", err)
+	}
+
+	return &UnixSocketConnector{
+		listener:   listener,
+		socketPath: hostSocket,
+		sandboxed:  sandboxed,
+		logger:     logger,
+	}, nil
+}
+
+// GetHostSocketPath returns the UNIX socket path on the host.
+func (us *UnixSocketConnector) GetHostSocketPath() string {
+	return us.socketPath
+}
+
+// GetGuestSocketPath returns the UNIX socket path on the guest.
+func (us *UnixSocketConnector) GetGuestSocketPath() string {
+	if !us.sandboxed {
+		return us.socketPath
+	}
+	return "/host.sock"
+}
+
+// ConfigureSandbox configures the process sandbox if needed.
+func (us *UnixSocketConnector) ConfigureSandbox(cfg *process.Config) error {
+	if !us.sandboxed {
+		return nil
+	}
+	cfg.BindRW[us.GetHostSocketPath()] = us.GetGuestSocketPath()
+	return nil
+}
+
+// Connect establishes a connection to the runtime.
+func (us *UnixSocketConnector) Connect(p process.Process) (net.Conn, error) {
+	if us.listener == nil {
+		return nil, fmt.Errorf("not initialized")
+	}
+
+	// Spawn goroutine that waits for a connection to be established.
+	connCh := make(chan interface{})
+	go func() {
+		lerr := us.listener.SetDeadline(time.Now().Add(runtimeConnectTimeout))
+		if lerr != nil {
+			connCh <- lerr
+			return
+		}
+		conn, lerr := us.listener.Accept()
+		if lerr != nil {
+			connCh <- lerr
+			return
+		}
+
+		connCh <- conn
+		close(connCh)
+	}()
+
+	var conn net.Conn
+	select {
+	case res := <-connCh:
+		// Got a connection or timed out while accepting a connection.
+		switch r := res.(type) {
+		case error:
+			return nil, fmt.Errorf("error while accepting runtime connection: %w", r)
+		case net.Conn:
+			conn = r
+		default:
+			panic("invalid type")
+		}
+	case <-p.Wait():
+		// Runtime has terminated before a connection was accepted.
+		us.logger.Debug("runtime process exited unexpectedly",
+			"pid", p.GetPID(),
+			"err", p.Error(),
+		)
+
+		return nil, fmt.Errorf("terminated while waiting for runtime to connect")
+	}
+	return conn, nil
+}
+
+// Close releases any resources associated with the connector.
+func (us *UnixSocketConnector) Close() {
+	if us.listener != nil {
+		us.listener.Close()
+		us.listener = nil
+	}
+	us.socketPath = ""
```

Comment:
> I would remove this line, always return configured path, and move creation of listener to Connect. This way, the constructor does not return an error and one can call Connect + Close multiple times.

The problem with this is that (for some provisioners) the socket needs to be available before the runtime process starts.

The general idea is that a connector instance is only for single use which is why the factory is there to create the connector. Otherwise an additional "Start" method would be needed which I don't think makes sense if there is already a factory that can set this up and the flow would always call "Start" (or similar) after construction.

Reply:
> The problem with this is that (for some provisioners) the socket needs to be available before the runtime process starts.

If I see correctly, we create a socket with factory, start the process (naked/bubblewrap), and connect in the end. An alternative approach would be to have a connector, then call connect which creates a sockets and tries to accept connections, and in the end start the process. This way, the connector wouldn't be depended on the process anymore, as `Connect(p process.Process)` could be changed to `Connect(ctx context.Context)` and the caller would decide when to abort connect, e.g. when the process fails.

I guess the (current) factory solution is also ok, although I hate when constructors return error or start something that cannot be controlled in tests (e.g. go routines).